utils.py

```diff
@@ -0,0 +1,51 @@
+import json
+
+import numpy as np
+import xarray as xr
+
+
+# https://stackoverflow.com/a/47626762
+class NumpyEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, np.ndarray):
+            return obj.tolist()
+        return json.JSONEncoder.default(self, obj)
+
+
+def extract_field_along_tracks(dataset, var_name, tracks):
+    """Extract data from a given field along one or more ship tracks.
+
+    Parameters
+    ----------
+    dataset : :class:`xarray.Dataset`
+        Dataset containing data to extract.
+    var_name : str
+        Name of the field (data variable).
+    tracks : dict-like
+        One or more ship tracks in the GeoJSON format (must be a
+        `FeatureCollection` with `LineString` features only).
+
+    Returns
+    -------
+    tracks_data : dict
+        Data extracted along ship tracks.
+
+    """
+    tracks_data = []
+
+    for feature in tracks["features"]:
+        points_lat, points_lon = [
+            xr.DataArray(list(points), dims="track")
+            for points in zip(*feature["geometry"]["coordinates"])
+        ]
+
+        da_result = dataset[var_name].sel(
+            lat=points_lat, lon=points_lon, method="nearest"
+        )
+
+        tracks_data.append({
+            "track_id": feature["properties"]["track_id"],
+            var_name: da_result.values
+        })
```

Comment:
I'd directly try and return GEOJson via Geopandas here: https://geopandas.org/gallery/create_geopandas_from_pandas.html#from-longitudes-and-latitudes

```suggestion
        df_result = da_result.to_dataset(name=var_name).to_dataframe()

        gdf_result = geopandas.GeoDataFrame(
            df_result, geometry=geopandas.points_from_xy(df_result.lon, df_result.lat))

        tracks_data.append({
            "track_id": feature["properties"]["track_id"],
            var_name: gdf_result.to_json()  # not sure if this is correct here...
        })
```

Reply:
The problem is that attaching values to the coordinates of a `LineString` feature is not supported by GeoJSON. I found a proposal [here](https://github.com/mapbox/geojson-coordinate-properties) but it doesn't seem very active.

Alternatively, we could accept a `FeatureCollection` of points (I think this is what returns `gdf_result.to_json()` in your suggestion), but then we lose the topology. I'm not sure which GeoJson model should be used to encode ship tracks, I guess it depends on some constraints in the frontend side. Whatever the model we choose, it should be the same in the request and the response.

Also, since the the coordinates are already provided in the query data (as GEOJson), I'm not sure if we actually need to return it. It might be good to keep small the amount of data transferred between server/client. `track_id` fields are returned in the response so that the values could easily be merged with the coordinates in the frontend application.

Reply:
It might be worth having a look a the [topojson](https://github.com/topojson/topojson) specs.

Reply:
Makes sense.